Makefile

```diff
@@ -16,265 +60,311 @@ SPACE :=
-$(PROTO_GEN_SRC): $(PROTO_FILES) go.mod
+# protoc generates everything in one shot, so we don't need thrift's per-target stuff.
+$(BUILD)/protoc: $(PROTO_FILES) $(BIN)/$(PROTOC_VERSION_BIN) $(BIN)/protoc-gen-go $(BIN)/protoc-gen-go-grpc | $(BUILD)
 	@mkdir -p $(PROTO_OUT)
-	$(BIN)/protoc \
+	@echo "protoc..."
+	@$(BIN)/$(PROTOC_VERSION_BIN) \
 		--plugin $(BIN)/protoc-gen-go \
 		--plugin $(BIN)/protoc-gen-go-grpc \
 		-I=$(PROTO_ROOT)/public \
 		-I=$(PROTO_ROOT)/internal \
-		-I=$(BIN)/protoc-zip/include \
+		-I=$(PROTOC_UNZIP_DIR)/include \
 		--go_out=. \
 		--go_opt=module=$(PROJECT_ROOT) \
 		--go-grpc_out=. \
 		--go-grpc_opt=module=$(PROJECT_ROOT) \
-		$$(find $(PROTO_DIRS) -name '*.proto')
-
-proto-compile: $(PROTO_GEN_SRC)
+		$(PROTO_FILES)
+	@touch $@
+
+# ====================================
+# Rule-breaking targets intended ONLY for special cases with no good alternatives.
+# ====================================
+
+.PHONY: .fake-codegen .fake-protoc .fake-thriftrw
+
+# buildkite / release-only target to avoid building / running codegen tools (protoc is unable to be run on alpine).
+# this will ensure that committed code will be used rather than re-generating.
+# must be manually run before (nearly) any other targets.
+.fake-codegen: .fake-protoc .fake-thrift
+
+# "build" fake binaries, and touch the book-keeping files, so Make thinks codegen has been run.
+# order matters, as e.g. a $(BIN) newer than a $(BUILD) implies Make should run the $(BIN).
+.fake-protoc: | $(BIN) $(BUILD)
+	touch $(BIN)/$(PROTOC_VERSION_BIN) $(BIN)/protoc-gen-go $(BIN)/protoc-gen-go-grpc
+	touch $(BUILD)/protoc
+
+.fake-thrift: | $(BIN) $(BUILD)
+	touch $(BIN)/thriftrw $(BIN)/thriftrw-plugin-yarpc
+	touch $(THRIFT_GEN)
+
+# ====================================
+# other intermediates
+# ====================================
+
+$(BUILD)/proto-lint: $(PROTO_FILES) $(BIN)/$(BUF_VERSION_BIN) | $(BUILD)
+	@cd $(PROTO_ROOT) && ../$(BIN)/$(BUF_VERSION_BIN) lint
+	@touch $@
+
+# note that LINT_SRC is fairly fake as a prerequisite.
+# it's a coarse "you probably don't need to re-lint" filter, nothing more.
+$(BUILD)/lint: $(LINT_SRC) $(BIN)/revive | $(BUILD)
+	@echo "lint..."
+	@$(BIN)/revive -config revive.toml -exclude './canary/...' -exclude './vendor/...' -formatter unix ./... | sort
+	@touch $@
+
+# fmt and copyright are mutually cyclic with their inputs, so if a copyright header is modified:
+# - copyright -> makes changes
+# - fmt sees changes -> makes changes
+# - now copyright thinks it needs to run again (but does nothing)
+# - which means fmt needs to run again (but does nothing)
+# and now after two passes it's finally stable, because they stopped making changes.
+#
+# this is not fatal, we can just run 2x.
+# to be fancier though, we can detect when *both* are run, and re-touch the book-keeping files to prevent the second run.
+# this STRICTLY REQUIRES that `copyright` and `fmt` are mututally stable, and that copyright runs before fmt.
```

Comment:
typo: mutu**t**ally

Reply:
bleh.  I should hunt down a spellchecker extension.  kinda surprised vscode doesn't come with one tbh, or use the OS's.